### DISCUSSION TITLE
Linebreaks after say, whisper, and emotes

### DISCUSSION THREAD

**Discussion by jungsunm** [evennia/evennia#2424] [Evennia Questions] [answered]
Hi, I'm just learning Evennia over the last few days, and I've set up a barebones server that I'm planning on using as a creative writing assignment (I'm a HS English teacher by trade). I didn't see this question already asked, perhaps because it's so basic, so my apologies in advance.

I'm looking to insert linebreaks before and after **says**, **whispers**, and **emotes** to improve readability, but can't find where to do it. I've looked in general.py and at_say, at_before_say, but the best I could do is something like this:

> An old wizard says, "
> 4 8 15 16 23 42
> "

where the linebreaks are before/after the self.args rather than the entire line.

Many thanks!

**Comment by davewiththenicehat** [comment, marked as answer]
The absolute easiest thing I can think to suggest. This method will only put a new line before your speech. Skip this if you want a new line before and after.
All objects that inherit `DefaultObject`, have a method called `at_before_say`. `at_before_say` is in `evennia.objects.objects.DefaultObject`.
Override `at_before_say` your `typeclasses.characters.Character` class. Replace the one active line in `at_before_say` with two
```python
return message
```
will now be
```python
self.msg(' ')
return message
```

If you would like new lines both before and after speech. This one requires a bit of typeclassing knowledge.
All objects that inherit `DefaultObject`, have a method called `at_say`. `at_say` is in `evennia.objects.objects.DefaultObject`.
Override `at_say` in your `typeclasses.characters.Character` class. You will want to make certain to pass all arguments. Than use python's `super` function to call the original `at_say` method.
I have not tested the code below but when done it should look itenditcal or likely very simular to:
```python
def at_say(
        self,
        message,
        msg_self=None,
        msg_location=None,
        receivers=None,
        msg_receivers=None,
        **kwargs,
    ):
    self.msg(' ')
    super().at_say(message, msg_self, msg_location, receivers, msg_receivers, **kwargs)
    self.msg(' ')
```

Hope that helps out.

**Reply by jungsunm** [reply]
Thanks for the info and the detailed explanation. I tried both suggestions, but neither seems to work. For the override of either at_before_say or at_say, you said to put it in the typeclasses.characters.Character class - I presume that's in the "mygame" directory and not the original?

**Reply by Griatch** [reply]
@jungsunm The solution by @davewiththenicehat works fine except an empty string will be stripped. You need to tell the server that a lone line break is really what you want by using the line break `\n` character. 

```python
# in mygame/typeclasses/characters.py

class Character(DefaultCharacter):
    # ...

    def at_say(self, *args, **kwargs):
        self.msg('\n')
        super().at_say(*args, **kwargs)
        self.msg('\n')        
```

> Note that if you want to insert line breaks from in-game (e.g. in descriptions using the `desc` command), you cannot use `\n` to do so (Evennia will strip such line characters for security reasons). You then need to use the string `'|/'`, which is the  Evennia tag for line break (see [text tags](https://www.evennia.com/docs/latest/TextTags.html#ansi-colours) for more info). You could also use `'|/'` above if you wanted.

> Second note: Text looking cramped can also be a client issue. A client that echoes the input back to the screen will often make text look cramped, for example. You can turn this off on the client side, which can make things look things look less cramped (e.g. in tintin++ you'll get an extra space between all commands this way).

**Comment by jungsunm** [comment]
Fantastic! Thanks @griatch and @davewiththenicehat - that works for me. I have a helper NPC that spits out tips at regular intervals, and it doesn't seem to affect it (it imports Character from typeclasses.characters) and uses `self.execute.cmd("say%s" % message)` that I got from the listening tutorial.

As far as other clients, I'll have the benefit that my students will all be using the webclient (as I am), so I have an unusual amount of control in that regard. On a side note, is there a way to echo the input and a status a la Diku in the webclient? It's off-topic, I know, and not something I need for this particular implementation.

Thanks again! I have just a couple more things to figure out before next week when I launch the project with my students (they won't be building w/i Evennia (this time around, at least)) :)

**Reply by Griatch** [reply]
@jungsunm Lots of questions in a small area there. :) 

- Echo is disabled in the web client by default. A user can turn it on in the client options (the cog wheel in the upper left corner). 
- If your NPC inherits from `Character`, then running `self.execute_cmd` on it should have it fire the same `msg` above. If not , put `prints` in your code (or run debugger in Evennia interactive mode if you are more experienced) to figure out just what is firing and where. 
 - With 'status', I presume you need a prompt? This is done with the `prompt` outputfunction. To test this, run the following in the web client (you need to be developer/superuser to use the `py` command):

        py self.msg(prompt="HP:10, SP:22, MP:12")

    Note: Don't forget to mark an answer when your question has been answered :)

**Reply by jungsunm** [reply]
Thanks (and for answering my other questions - sorry for packing them into an unrelated topic)! Yeah - I'm not sure why the linebreaks aren't happening with the NPC. I surrounded the linebreaks with `Print`s, which show they're firing, but... shrug. At this point, it's not a big deal, since it works with players.

Off to figure out a few more things!

**Reply by Griatch** [reply]
Actually, since you use `execute_cmd` (which mimics data being input from the user) you may need to put `|/` markers instead of `\n` for the reasons given earlier.

**Reply by jungsunm** [reply]
Yeah - I've tried both, with no visible effect. Is there an alternative to `execute_cmd`? Like I said, I'm willing to move on from this one. :) Thanks again!

**Reply by Griatch** [reply]
All `execute_cmd` does is execute a command class, which in turn is just a collection of evennia api calls. It's generally recommended to just do those api calls directly instead of going via the command-system when you have doing stuff from the code anyway. You can just as well have the NPC call `self.location.msg_contents(...)` yourself, in this case.

**Reply by jungsunm** [reply]
Ah - sorry - that seems like such a simple solution (one that I probably should have known were I more experienced with Evennia and Python), but that did the trick. Obviously I had to mimic the say, but otherwise, success!

> An old wizard says, "4 8 15 16 23 42"
>
> An old wizard says, "GET or TAKE items you see - they might be useful later!"

**Comment by kovitikus** [comment]
Screen spam and quick scrolling can be a problem in text games. Most players that have been in a text game long enough want to keep as much information available on the screen at once.

Another possible solution to your problem is to simply have new lines indented, giving a visual indication of where new output begins, separating it from the previous output.

Although, this isn't always helpful if you have multiple lines that are 1 line in length. It's something to think about though.

**Reply by MrWigggles** [reply]
@jungsunm Isnt making a text game.
They're making a shared classroom enviroment for writing.
If anything, I would go further and add in Line Counts to make it easier to get folks to look at the same line since they cant share a cursor.
Also Indents are different per client. Some clients have a very small indent, some have large one.

**Reply by jungsunm** [reply]
Thanks! I mean, _technically_ it is a game, just without the full Diku MUD experience I had when I first dreamed this up. It's a creative writing project where I'm giving students a chance to write a few rooms reflecting themselves and their own interest. 

They'll be able to walk around, read each others' areas, interact with each other, and perhaps do a scavenger hunt (whether built into the game or out remains to be seen). My district goes longer than most schools in the area, and I wanted to do something different and hopefully engaging that would work on developing their writing skills.

**Reply by MrWigggles** [reply]
Sounds great. 
https://podcasts.apple.com/us/podcast/tinytalk-episode-022-tales-of-a-fourth-grade-something/id250657072?i=1000062466412
That was a podcast of forth graders using Mush and finding the experience quite engaging. You can play on that web page. I know their forth graders and not high schoolers but I hope some of its useful. 
But yea, if you want to have folks be able to review and critque each other, I'd really suggest number the lines for the room description to make it easier to reference.
